src/payload.ts

```diff
@@ -1,195 +1,212 @@
-        break;
-      case PayloadType.PoisonMicroblock:
-        // TODO: implement
-        break;
-      case PayloadType.Coinbase:
-        this.coinbaseBuffer = bufferReader.read(COINBASE_BUFFER_LENGTH_BYTES);
-        break;
-      default:
-        throw new Error(
-          `Unexpected transaction payload type while deserializing: ${this.payloadType}`
-        );
-    }
+  if (typeof memo === 'string') {
+    memo = memoString(memo);
   }
+
+  return {
+    type: StacksMessageType.Payload,
+    payloadType: PayloadType.TokenTransfer,
+    recipientAddress,
+    amount,
+    memo: memo ? memo : memoString(''),
+  };
 }
 
-export class TokenTransferPayload extends Payload {
-  constructor(recipientAddress?: string, amount?: BigNum, memo?: string) {
-    super();
-    this.payloadType = PayloadType.TokenTransfer;
+export interface ContractCallPayload {
+  readonly type: StacksMessageType.Payload;
+  readonly payloadType: PayloadType.ContractCall;
+  readonly contractAddress: Address;
+  readonly contractName: LengthPrefixedString;
+  readonly functionName: LengthPrefixedString;
+  readonly functionArgs: ClarityValue[];
+}
 
-    this.recipientAddress = new Address(recipientAddress);
-    this.amount = amount;
-    this.memo = memo ? new MemoString(memo) : new MemoString('');
+export function contractCallPayload(
+  contractAddress: string | Address,
+  contractName: string | LengthPrefixedString,
+  functionName: string | LengthPrefixedString,
+  functionArgs: ClarityValue[]
+): ContractCallPayload {
+  if (typeof contractAddress === 'string') {
+    contractAddress = address(contractAddress);
+  }
+  if (typeof contractName === 'string') {
+    contractName = lengthPrefixedString(contractName);
+  }
+  if (typeof functionName === 'string') {
+    functionName = lengthPrefixedString(functionName);
   }
+
+  return {
+    type: StacksMessageType.Payload,
+    payloadType: PayloadType.ContractCall,
+    contractAddress,
+    contractName,
+    functionName,
+    functionArgs: functionArgs,
+  };
 }
 
-export class ContractCallPayload extends Payload {
-  constructor(
-    contractAddress?: string,
-    contractName?: string,
-    functionName?: string,
-    functionArgs?: ClarityValue[]
-  ) {
-    super();
-    this.payloadType = PayloadType.ContractCall;
-    this.contractAddress = new Address(contractAddress);
-    this.contractName = new LengthPrefixedString(contractName);
-    this.functionName = new LengthPrefixedString(functionName);
-    this.functionArgs = functionArgs;
+export interface SmartContractPayload {
+  readonly type: StacksMessageType.Payload;
+  readonly payloadType: PayloadType.SmartContract;
+  readonly contractName: LengthPrefixedString;
+  readonly codeBody: LengthPrefixedString;
+}
+
+export function smartContractPayload(
+  contractName: string | LengthPrefixedString,
+  codeBody: string | LengthPrefixedString
+): SmartContractPayload {
+  if (typeof contractName === 'string') {
```

Comment:
`isString` helper would be cool, seeing lots of `typeof`s

Reply:
making use of Typescript's `typeof` type guards feature here: https://www.typescriptlang.org/docs/handbook/advanced-types.html#typeof-type-guards

Reply:
They use a helper fn the example 

```
function isString(x: any): x is string {
    return typeof x === "string";
}
```

Reply:
yea but directly below they say:
> However, having to define a function to figure out if a type is a primitive is kind of a pain. Luckily, you don’t need to abstract typeof x === "number" into its own function because TypeScript will recognize it as a type guard on its own. That means we could just write these checks inline.